### ISSUE TITLE
Reference to a window after reloading is broken

### ISSUE THREAD

**Issue by franchuk** [nwjs/nw.js#4773]
In my test app i open a window on app start
`nw.Window.open(nw.App.manifest.onlineURL, windowProperties, function (win) {
	        win.on('close', function (shouldQuit) {
	            if (shouldQuit) {
	                nw.App.quit();
	            } else {
	                win.hide()
	            };
	        });
	        nw.App.on('reopen', function () {
	            win.show();
	        });
	        tray.on('click', function () {
	        	win.show();
	        });
            win.window.onload = function (){
                setTimeout(function(){
                    win.window.location.href = nw.App.manifest.onlineURL;
                }, 5000);
            }
	    });`
And i have a Tray click handler.
After reloading the window tray.on('click) does not work. The window is hidden and i do not have any reference to it.

NW 0.14.2 Windows 10 x64.
My test app:
[test_app.zip](https://github.com/nwjs/nw.js/files/236500/test_app.zip)

**Comment by franchuk** [comment]
any updates on this?

**Comment by rogerwang** [comment]
@ghostoy has the answer for this like he did in many other similar issues

**Comment by RohitSharma27** [comment]
Can you please provide more information like when does the file that contains the above code is executed? When do you inject the file that contains your tray's code? 
I did face the same issue and the reason is when you reload or navigate, your functions and variables are likely to get GCed. So you are gonna have to write your tray related code in background context, which stays there for your app's lifetime. See manifest format and use bgscript in your package.json to include your tray file. I hope it helps. Cheers. :)

**Comment by ghostoy** [comment]
@RohitSharma27 :+1: 
It's recommended to setup menus and trays in background page to prevent being GCed.

**Comment by davidcampos** [comment]
@franchuk Have you managed to fix this?
I create the window in the init script (main: index.js on manifest) and keep the reference to it. After navigating to another page, the window reference is broken and the close event is cleared.
I have tray with menu on the background page.

@ghostoy and @rogerwang Any recommendations?

Thanks for your help.

**Comment by ghostoy** [comment]
@davidcampos Could you post the code how you use the tray?

**Comment by RohitSharma27** [comment]
@davidcampos where are you keeping the window's reference? And, when you navigate to other page, do you mean to open it in the same window? If this is the case then, your window's previous content is going to get GCed

**Comment by davidcampos** [comment]
@ghostoy On "index.js", I keep a variable of the "TrayManager" (code below). Is this OK or I should do it in a different manner?
```
function TrayManager(windowManager) {
    // Give it a menu
    this.menu = new nw.Menu();
    this.menu.append(new nw.MenuItem({
        type: 'normal',
        label: 'Quit',
        click: function () {
            nw.App.quit();
        }
    }));

    // Create a tray icon
    this.tray = new nw.Tray({icon: 'tray.png', tooltop: 'PACScenter', menu: this.menu});
}

module.exports = TrayManager;
```


@RohitSharma27 On "index.js" I create a windowManager and keep the window references there. I open a login URL on the window, after performing login and redirecting to another page on the same window, I loose the window reference. I am doing something wrong?

**Comment by RohitSharma27** [comment]
See, the 'index.js' file that you have included as main : 'index.js' via package.json in your app is going to be in the background context. So, the variables in it are not going to get lost unless you close your app. Page navigation or reload only refreshes the DOM context. The code that you have posted above is going to be included in your index.js via require('fileName'), I assume. So, it's not going to get GCed either. It would be helpful if you can post a minimal sample app. Cheers! :)

**Comment by davidcampos** [comment]
@RohitSharma27 

> The code that you have posted above is going to be included in your index.js via require('fileName'), I assume. 

Yes, the Window manager is included on index via require, which keeps references to Windows. However, by just navigating to a different page on the primary Window, by using the reference I am no longer able to call show, hide or close methods. Moreover, the close event previously set no longer works.
However, if I store the DOM window (Window.window attribute), even after navigating to a different page, I am able to access all DOM elements.

That is why I do not understand why the Window reference no longer "works" after navigating to a different page in the same window.

> By the way, which NW version are you using?

I am using NW 0.16.1.

**Comment by davidcampos** [comment]
I found a way to fix this.
Every time I reload the page, I renew the Window reference by using:
`this.primaryWindow = nw.Window.get(this.primaryWindowDOM);`

I also had to set the close, minimize and restore events again.
I hope this helps other people.
Thanks for your help.

**Comment by rogerwang** [comment]
should be fixed in #5294 and #5312

**Comment by lxhao** [comment]
I solved this problem, the problem is  caused by reassignment to variable win after reloading page . solve method  please see my code https://github.com/lxhao/nwjsReload.git
